File: ios/Plugin/AwsAmplifyPlugin.swift
import Foundation
import Amplify
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(AwsAmplifyPlugin)
public class AwsAmplifyPlugin: CAPPlugin {
    private let implementation = AwsAmplify()
    
    @objc func load(_ call: CAPPluginCall) {
        let cognitoConfig = call.getObject("cognitoConfig")!
        
        implementation.load(
            cognitoConfig: cognitoConfig,
            onSuccess: {
                call.resolve()
            },
            onError: {error in
                print(error)
                call.reject(error.localizedDescription)
            })
    }
    
    @objc func signIn(_ call: CAPPluginCall) {
        let email = call.getString("email") ?? ""
        let password = call.getString("password") ?? ""
        
        implementation.signIn(
            email: email,
            password: password,
            onSuccess: {data in
                self.implementation.fetchAuthSession(
                    onSuccess: {session in
                        call.resolve(session)
                    },
                    onError: {error in
                        call.reject(error.localizedDescription)
                    })
            },
            onError: {error in
                print(error)
                call.reject(error.localizedDescription)
            })
    }
    
    @objc func signOut(_ call: CAPPluginCall) {
        self.implementation.signOut(
            onSuccess: { response in
                call.resolve(response)
            }, onError: { error in
                call.reject(error.localizedDescription)
            })
    }
    
    @objc func federatedSignIn(_ call: CAPPluginCall) {
        let provider = call.getString("provider") ?? ""
        
        self.implementation.fetchAuthSession(
            onSuccess: {session in
                if(session["status"] as! Int == 0) {
                    call.resolve(session)
                } else {
                    self.implementation.federatedSignIn(
                        provider: provider,
                        onSuccess: { result in
                            if(result["status"] as! Int == 0) {
                                self.fetchAuthSession(call)
                            } else {
                                call.resolve(result)
                            }
                        },
                        onError: { error in
                            print(error)
                            call.reject(error.localizedDescription)
                        })
                }
                
            },
            onError: {error in
                call.reject(error.localizedDescription)
            })
        
        
    }
    @objc func fetchAuthSession(_ call: CAPPluginCall) {
        
        self.implementation.fetchAuthSession(
            onSuccess: {session in
                call.resolve(session)
            },
            onError: {error in
                call.reject(error.localizedDescription)
            })
    }
    
    @objc func getUserAttributes(_ call: CAPPluginCall) {
        
        self.implementation.getUserAttributes(
            onSuccess: {session in
                call.resolve(session)
            },
            onError: {error in
                call.reject(error.localizedDescription)
            })
    }
    
    @objc func updateUserAttributes(_ call: CAPPluginCall) {
        
        if let userAttributes = call.getArray("attributes") as JSArray?{
            self.implementation.updateUserAttributes(
                userAttributes: userAttributes.map {
                    let attribute = $0 as! JSObject
                    return AuthUserAttribute(getAttributeKey(key: attribute["name"] as! String), value: attribute["value"] as! String)
                },
                onSuccess: {session in
                    call.resolve(session)
                },
                onError: {error in
                    call.reject(error.localizedDescription)
                })
        }
        
    }
    
    @objc func deleteUser(_ call: CAPPluginCall) {
        implementation.deleteUser(
            onSuccess: {
                call.resolve(["status": 0])
            },
            onError: {error in
                call.resolve(["status": -1])
            })
    }
    
    private func getAttributeKey(key: String) -> AuthUserAttributeKey{
        switch(key){
        case "address":
            return AuthUserAttributeKey.address
        case "birthDate":
            return AuthUserAttributeKey.birthDate
        case "email":
            return AuthUserAttributeKey.email
        case "familyName":
            return AuthUserAttributeKey.familyName
        case "gender":
            return AuthUserAttributeKey.gender
        case "givenName":
            return AuthUserAttributeKey.givenName
        case "locale":
            return AuthUserAttributeKey.locale
        case "middleName":
            return AuthUserAttributeKey.middleName
        case "name":
            return AuthUserAttributeKey.name
        case "nickname":
            return AuthUserAttributeKey.nickname
        case "phoneNumber":
            return AuthUserAttributeKey.phoneNumber
        case "picture":
            return AuthUserAttributeKey.picture
        case "preferredUsername":
            return AuthUserAttributeKey.preferredUsername
        default:
            return AuthUserAttributeKey.custom(key)
        }
    }
    
    // var permissionCallID: String?
    // var locationManager: CLLocationManager?
    
    // @objc override public func requestPermissions(_ call: CAPPluginCall) {
    //     if let manager = locationManager, CLLocationManager.locationServicesEnabled() {
    //         if CLLocationManager.authorizationStatus() == .notDetermined {
    //             bridge?.saveCall(call)
    //             permissionCallID = call.callbackId
    //             manager.requestWhenInUseAuthorization()
    //         } else {
    //             checkPermissions(call)
    //         }
    //     } else {
    //         call.reject("Location services are disabled")
    //     }
    // }
}

File: android/src/main/java/com/capacitor/aws/amplify/AwsAmplifyPlugin.java
package com.capacitor.aws.amplify;

import android.os.Build;

import androidx.annotation.RequiresApi;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "AwsAmplify")
public class AwsAmplifyPlugin extends Plugin {

    private final AwsAmplify implementation = new AwsAmplify();

    @RequiresApi(api = Build.VERSION_CODES.N)
    @PluginMethod
    public void load(PluginCall call) {
        var cognitoConfig = call.getObject("cognitoConfig");

        implementation.load(
                cognitoConfig,
                bridge.getContext(),
                result -> call.resolve(),
                error -> {
                    call.reject(error.toString());
                });
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @PluginMethod
    public void signOut(PluginCall call) {

        implementation.signOut(
                call::resolve,
                error -> {
                    call.reject(error.toString());
                });
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @PluginMethod
    public void federatedSignIn(PluginCall call) {
        String provider = call.getString("provider");

        implementation.federatedSignIn(
                provider,
                this.getActivity(),
                call::resolve,
                error -> {
                    call.reject(error.toString());
                });
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @PluginMethod
    public void fetchAuthSession(PluginCall call) {

        implementation.fetchAuthSession(
                call::resolve,
                error -> {
                    call.reject(error.toString());
                });
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @PluginMethod
    public void getUserAttributes(PluginCall call) {

        implementation.getUserAttributes(
                call::resolve,
                error -> {
                    call.reject(error.toString());
                });
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @PluginMethod
    public void updateUserAttributes(PluginCall call) {
        var userAttributes = call.getArray("attributes");
        implementation.updateUserAttributes(userAttributes,
                call::resolve,
                error -> {
                    call.reject(error.toString());
                });
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @PluginMethod
    public void deleteUser(PluginCall call) {

        implementation.deleteUser(
                call::resolve,
                error -> {
                    call.reject(error.toString());
                });
    }

}
